packages/apollo-gateway/src/make-fetch-happen.d.ts

```diff
@@ -0,0 +1,40 @@
+declare module 'make-fetch-happen' {
+  import { Response, Request } from 'apollo-server-env';
+
+  // If adding to these options, they should mirror those from `make-fetch-happen`
+  // @see: https://github.com/npm/make-fetch-happen/#extra-options
+  export interface FetcherOptions {
+    cacheManager?: string | CacheManager;
+    // @see: https://www.npmjs.com/package/retry#retrytimeoutsoptions
+    retry?:
+      | boolean
+      | number
+      | {
+          //  The maximum amount of times to retry the operation. Default is 10. Seting this to 1 means do it once, then retry it once
+          retries?: number;
+          // The exponential factor to use. Default is 2.
+          factor?: number;
+          // The number of milliseconds before starting the first retry. Default is 1000.
+          minTimeout?: number;
+          // The maximum number of milliseconds between two retries. Default is Infinity.
+          maxTimeout?: number;
+          // Randomizes the timeouts by multiplying with a factor between 1 to 2. Default is false.
+          randomize?: boolean;
+        };
+    onRetry(): void;
+  }
+
+  export interface CacheManager {
+    delete(req: Request): Promise<Boolean>;
+    put(req: Request, res: Response): Promise<Response>;
+    match(req: Request): Promise<Response | undefined>;
+  }
+  export interface Fetcher {
+    (url: string): Promise<Response>;
```

Comment:
The `fetch` that `make-fetch-happen` still accepts a `Request`  (in place of `string` for the first argument), but also options passed as the second argument (known as "init", by [the specification](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch) which are represented by the `RequestInit` type.  That type lives on the [`dom` lib in TypeScript](https://github.com/Microsoft/TypeScript/blob/bab0c995849eff9d5617653de537b7f9103c7e10/lib/lib.dom.d.ts#L1549-L1602) (warning, large file), but we intentionally have a rough-copy of it in our `apollo-server-env`:

https://github.com/apollographql/apollo-server/blob/bbe3dee5bce50f39459b92192edf3b58b89125fb/packages/apollo-server-env/dist/fetch.d.ts#L47-L71

Additionally, I think that the default options that `fetch` accepts in `RequestInit` are extended with those supported by `make-fetch-happen` — represented (but I haven't validated the completeness of it) by `FetcherOptions` above.

I think we might want something like this:

```typescript
  import { fetch as envFetch } from 'apollo-server-env';
  export interface Fetcher {
    (
      input?: Parameters<typeof envFetch>[0],
      init?: Parameters<typeof envFetch>[1] & FetcherOptions,
    ): ReturnType<typeof envFetch>;
  }
```

Reply:
From a code readability standpoint, I chose to use the types themselves and effectively dupe the `fetch` type here: 2b3f234

I won't deny that your approach couples the types and some potential maintenance burden of duplicated types. Interested to know your thoughts, since I'm pretty torn, tbh.

Reply:
I guess my perspective is exactly that.  In other words, I suggested this because I don't really care to keep myself up to date on what new options the Fetch API eventually introduces; there was a new draft of the specification that landed in January.  The, e.g., `node-fetch` package is likely going to stay on top of that, and I'm glad they will.

However, my suggestion here doesn't solve an existing problem — we already maintain our own `Fetch` types in `apollo-server-env`. 😬   More than anything, I wish _that_ were not the case right now and my biggest concern is that we continue in that direction  — we already have `apollo-env` and `apollo-server-env` in some of the `@apollo/*` packages.  

Let's certainly ship this how it is for now, but just keep our eyes on it.